### ISSUE TITLE
[BUG] Where to get the refresh_token and access_token

### ISSUE THREAD

**Issue by franva** [KenEucker/imgur#223]
**Describe the bug**
I'm trying to upload images to my own album. 

**To Reproduce**
Here is the details with code:
https://stackoverflow.com/questions/77889717/how-to-add-an-uploaded-image-to-my-own-imgur-album

**Expected behavior**
I saw the example code here 
https://github.com/KenEucker/imgur/blob/main/example/index.html

But it does not say where to get these 2 important tokens.

**Screenshots**

**Environment (please complete the following information):**
 - Library version [e.g. 1.0.0]:
 - OS [e.g. macOS]:
 - NodeJS version [e.g.14.16.0]:

**Additional context**

**Comment by franva** [comment]
I got the answer and found that I need to call the `authorize` ImgUr endpoint with client id and client secret to get access token and refresh token.

**Comment by KenEucker** [comment]
I was going to check in about that. 

Anonymous uploads should work without requiring an access token, but, in order to upload to an album you have access to you need to use an accessToken.

Cheers!

**Comment by franva** [comment]
> I was going to check in about that.
> 
> Anonymous uploads should work without requiring an access token, but, in order to upload to an album you have access to you need to use an accessToken.
> 
> Cheers!

Thanks Ken :)

I have figured out where to get these 2 tokens for auth approach.

For un-auth approach, it requires the `deleteHash` to add images to an album.
But I am not sure where to find out the `deleteHash` to be honest.
Appreciated if you could enlighten me on this.

Best regards,

Winston

**Comment by KenEucker** [comment]
@franva the delete hash is in the album data:
https://api.imgur.com/models/album

```
const album = imgur.getAlbum('xbvhXo')
console.log(album.deletehash)
```

**Comment by franva** [comment]
> @franva the delete hash is in the album data: https://api.imgur.com/models/album
> 
> ```
> const album = imgur.getAlbum('xbvhXo')
> console.log(album.deletehash)
> ```

I did that, and the `deletehash` contains nothing. 
But that could because I didn't put in the access_token. 

Let me try again and will update it here.

**Comment by KenEucker** [comment]
In my testing I am also seeing that the deletehash is not getting set. I think that Imgur isn't sending it in the response. It says OPTIONAL in the docs, but I don't know what makes it optional.

**Comment by franva** [comment]
Hi @KenEucker 

I can confirm that it comes back with all the data including the `deletehash` 

![image](https://github.com/KenEucker/imgur/assets/5151146/477dfeca-50ba-4d0f-bb2c-33b34a2aa0f1)

Did you put the accessToken when initializing client object?

I included it, I guess this is the reason.

**Comment by KenEucker** [comment]
@franva that's great. I guess in my testing I didn't realize you need to add both the refreshToken AND accessToken, as it didn't work with either individually.

Would you find it helpful it there was a method for retrieving tokens from Imgur, instead of needing to do that yourself? 

(This is supposed to be done for you, internally, via a method called getAuthenticationHeader, which is called on every request through axios. I can also look into if there is a bug there).

**Comment by franva** [comment]
@KenEucker 
I definitely think it is a big PLUS if the npm package can do this for me automatically.

I mean, I do think it would be a great feature to add in. 

The current procedure for authorizing access of ImgUr is: 

1. I need to call 
https://api.imgur.com/oauth2/authorize?client_id=<myclientid>&response_type=token&state=<anything> it myself.
2. This is okay, but then this api call opens a webpage and it requires us to manually click an `Allow` button to authorize the access, then it redirects us to the callback url.

This is a huge inconvenience, as I use the ImgUr api to store many images on my website, and the users of my website can also upload images there. I definitely do not want my users out of sudden to see a popup window showing the ImgUr Authorization page when they are uploading images.

If the new functionality you are going to add in can solve this, it would be a great feature for many people like me.

So what's your plan?

**Comment by KenEucker** [comment]
@franva I think you misread my comment or didn't see the part that was supposed to be strikethrough.

There is no workaround for this, as it is an OAUTH2 process that must be completed with a logged-in Imgur user. You can log into Imgur yourself, grant access rights to the application, gaining access to a refresh and access token, and then use those tokens to add/remove images from your personal Imgur account. This is something you only need to do once if you retain the refresh tokens.

However, if you intend to have users on your website use their own personal Imgur accounts, they will need to complete this step in this way. There is no other way to get access and refresh tokens.

More pertinent information from Imgur's documentation:

> **Stuff not to do**
> If someone else might own the copyright to it, don't upload it. Don't upload gore, "hate speech" (i.e. demeaning race, gender, age, religious or sexual orientation, etc.), or material that is threatening, harassing, defamatory, or that encourages violence or crime. Don't upload illegal content such as child sexual abuse material or nonconsensual ("revenge") porn. Don't hotlink to adult content or to file-sharing, gambling, torrent, warez, or Imgur rip-off sites. Don't impersonate someone else. **Also, don't use Imgur to host image libraries you link to from elsewhere, content for your website, advertising, avatars, or anything else that turns us into your content delivery network.** If you do – and we will be the judge – or if you do anything illegal, in addition to any other legal rights we may have, we will ban you along with the site you're hotlinking from, delete all your images, report you to the authorities if necessary, and prevent you from viewing any images hosted on Imgur.com. We mean it.

**Comment by franva** [comment]
> @franva I think you misread my comment or didn't see the part that was supposed to be strikethrough.
> 
> There is no workaround for this, as it is an OAUTH2 process that must be completed with a logged-in Imgur user. You can log into Imgur yourself, grant access rights to the application, gaining access to a refresh and access token, and then use those tokens to add/remove images from your personal Imgur account. This is something you only need to do once if you retain the refresh tokens.
> 
> However, if you intend to have users on your website use their own personal Imgur accounts, they will need to complete this step in this way. There is no other way to get access and refresh tokens.
> 
> More pertinent information from [Imgur's terms of service](https://imgur.com/tos):
> 
> > **Stuff not to do**
> > If someone else might own the copyright to it, don't upload it. Don't upload gore, "hate speech" (i.e. demeaning race, gender, age, religious or sexual orientation, etc.), or material that is threatening, harassing, defamatory, or that encourages violence or crime. Don't upload illegal content such as child sexual abuse material or nonconsensual ("revenge") porn. Don't hotlink to adult content or to file-sharing, gambling, torrent, warez, or Imgur rip-off sites. Don't impersonate someone else. **Also, don't use Imgur to host image libraries you link to from elsewhere, content for your website, advertising, avatars, or anything else that turns us into your content delivery network.** If you do – and we will be the judge – or if you do anything illegal, in addition to any other legal rights we may have, we will ban you along with the site you're hotlinking from, delete all your images, report you to the authorities if necessary, and prevent you from viewing any images hosted on Imgur.com. We mean it.

Hi @KenEucker ,

Thanks for your update.

I guess I didn't explain it clear. I will try it again:

The website does not require its users client id and client secret of ImgUr.
The website uses my credential of ImgUr.

Also, thanks for showing the ImgUr's terms. It looks like I will need to find something else.
Appreciated if you could share some other websites or services which provide free image storage.

Back to your idea(now stroked through), I still think it will be useful.

Cheers,

Winston
